### ISSUE TITLE
Add rich reprs to STAC objects

### ISSUE THREAD

**Issue by TomAugspurger** [stac-utils/pystac#508]
IPython / jupyter define a protocol where objects defining methods like `_repr_html_` will have the ability to produce rich representations in a notebook. See https://ipython.readthedocs.io/en/stable/config/integrating.html.

I think this would be very useful for exploring STAC objects. https://github.com/dask/distributed/pull/4857 has some designs / code that might be worth stealing. In particular the handling of nested objects would apply to catalogs, collections, and even items.

See http://matthewrocklin.com/blog/2019/07/04/html-repr for some background.

**Comment by DahnJ** [comment]
This seems like an immensely useful feature to have, especially given the nested structure STAC has.

PR https://github.com/stac-utils/pystac/pull/573 was closed since it used an extra dependency `jinja2`. The agreed-upon solution was to create a new package `pystac-jupyter`. Did that ever get made? If not, what would be needed to help it get made?

In case it didn't get made: would it perhaps be possible/convenient to only include `jinja2` as an [optional dependency](https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#optional-dependencies) and only attach/display the HTML representations if `jinja2` is installed? I haven't thought much about this, just thought I'd mention it.

cc: @moradology

**Comment by TomAugspurger** [comment]
IMO, this would have to go in pystac itself to be worthwhile. https://ipython.readthedocs.io/en/stable/config/integrating.html#formatters-for-third-party-types mentions a way for third-party libraries to register formatters for types in other libraries, but that's not too user-friendly.

I think an optional dependency is a good compromise. The `_repr_html_` would then just need a

```python
try:
    import jinja2
except ImportError:
    return repr(self)
...
```

And in case it's helpful, `rich.inspect` can give a (verbose) repr:

<img width="754" alt="image" src="https://user-images.githubusercontent.com/1312546/144679292-9324d527-f11a-47bc-9900-4502291de733.png">

But I think that the nested nature of STAC objects means things like HTML dropdowns / disclosure triangles will be helpful.

**Comment by duckontheweb** [comment]
Closed by #743
